### PR TITLE
Areas of D usage

### PR DESCRIPTION
**Pull request by wilzbach**
As the _first five minutes matter_, I thought it would be nice if we have an overview of areas in which D is used.
It is different to the great [Overview](http://dlang.org/overview.html) page as it tries for every area to inform about
-  applications written in D
-  key reasons/selling points

That being said, this document is a draft and your feedback & review is highlighy appreciated.
If I left out an area or you miss a great application - please let me know!

Also note that I propose to put this on dlang.org, because in contrast to the wiki we can go through the very helpful peer-review stage and have the official branding.

### CODE REVIEW

**Comment by wilzbach** [comment]
Can someone have a look at the content?
I beliebe providing such an overview easily accessible from the front page is part of our mission to make D more mainstream!

**Comment by schuetzm** [comment]
Agreed that this is important. I wouldn't put "Bare metal" at the top of the list, though. It's not D's strongest point currently. Also, "Compilers" are not really bare metal applications.

File: areas-of-d-usage.dd [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+$(SECTION2 $(LNAME2 industry, Industry),
+
+$(SECTION3 $(LNAME2 games, Games),
+
+    Binary compatibility to existing libraries is very important in large-scale
+    projects. $(BR)
+    Remedy Entertainment has successfully shipped the first $(HTTP www.quantumbreak.com/,
+    AAA game) to use D code for XBox One and Windows 10. $(BR)
+    For more details see Ethan Watson's
+    $(HTTP http://dconf.org/2016/talks/watson.html, presentation).
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: The `http://` prefix mustn't be used with `HTTP`.

File: areas-of-d-usage.dd [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    $(HTTPS github.com/d-gamedev-team/gfm, gfm).
+)
+
+$(SECTION3 $(LNAME2 web, Web applications),
+
+    D's ability to combine low and high-level code in one language, asynchronous
+    features like Fibers and compile-time function execution makes it well suited
+    for high-performance asynchronous I/O, concurrency and web applications like
+    $(HTTPS https://github.com/rejectedsoftware/vibe.d, vibe.d). $(BR)
+
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: ditto

File: areas-of-d-usage.dd [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    As an example, have a look at this talk about a
+    $(HTTPS https://www.youtube.com/watch?v=_gCFpHxAJRc, Mobile Gameserver Backend in D)
+)
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: ditto

File: areas-of-d-usage.dd [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+    One of the most popular packages for numerical computing is NumPy. It has a
+    great feature set, easy API and is fast due to C bindings. However for every
+    feature that is not in NumPy works a very low speed.
+    D allows to write high-performant code for
+    $(HTTP github.com/libmir/mir, numerical computations) that runs as efficient
+    as they would have been written in C.
+    Additionally due to D's introspection at compile-time it enables
+    further optimizations for superior performance. See Andrei's talk
+    $(HTTPS https://www.youtube.com/watch?v=AxnotgLql0k, Fastware) on this topic.
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: ditto

File: areas-of-d-usage.dd [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    with more additions to come in the future.
+)
+
+$(SECTION3 $(LNAME2 gpu, GPU Programming),
+
+    For most programmers used to developing for a modern CPU,
+    working with GPUs for general purpose computation can appear daunting and laborious.
+    Even seasoned users will admit that the performance improvements are often not worth the extra effort. $(BR)
+    Thanks to the power the power of D’s type system and static introspection
+    $(HTTPS http://dconf.org/2016/talks/colvin.html, GPU Programming) is
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: Don't use `https://`

File: areas-of-d-usage.dd [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    great feature set, easy API and is fast due to C bindings. However for every
+    feature that is not in NumPy works a very low speed.
+    D allows to write high-performant code for
+    $(HTTP github.com/libmir/mir, numerical computations) that runs as efficient
+    as they would have been written in C.
+    Additionally due to D's introspection at compile-time it enables
+    further optimizations for superior performance. See Andrei's talk
+    $(HTTPS https://www.youtube.com/watch?v=AxnotgLql0k, Fastware) on this topic.
+
+    D's standard library includes support for $(HTTPS $(ROOT_DIR)phobos/std_experimental_ndslice.html, multidimensional arrays)
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: This results in a broken link, see Vladimir's auto tester.

File: areas-of-d-usage.dd [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    $(HTTP tech.adroll.com/blog/data/2014/11/17/d-is-for-data-science.html, "D is for Data Science")
+    which explains why D is a keystone language for their critical infrastructure.
+)
+
+$(SECTION3 $(LNAME2 teaching, Teaching),
+
+    D's multi-paradigm (imperative, structured, object oriented, generic, functional programming
+    purity, and even assembly) approach allows to teach in one language and gradually
+    explain new features without needing to switch to a different language.
+    D also has a memory-safe subset $(I $(WEB ./safed.html, SafeD)). It's safety and
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: This link is broken, too, see the auto tester.

**Comment by wilzbach** [comment]
> Agreed that this is important. I wouldn't put "Bare metal" at the top of the list, though.

How would you structure it?

> It's not D's strongest point currently. Also, "Compilers" are not really bare metal applications.

Renamed the category to System programming, fixed all the links and went over the text again.

File: areas-of-d-usage.dd [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            $(LI $(RELATIVE_LINK2 teaching, Teaching))
+        )
+    )
+)
+
+This list highlights a selection of great applications written in D
+from various areas. $(BR)
+
+For an overview of special features of D, have a look at the
+$(WEB $(ROOT_DIR)overview.html, D Overview page).
```
Comment (schuetzm) [inline areas-of-d-usage.dd]: This one is still broken, need to use `LINK2` as below...

**Comment by schuetzm** [comment]
I'd say by "popularity" of the usage area. For example "Numerical computing / Data science" seems to be a hot topic, judging from the forum activity. On the other hand, "Game development" is sexy. But we probably shouldn't make a science out of it.

**Comment by wilzbach** [comment]
> I'd say by "popularity" of the usage area. For example "Numerical computing / Data science" seems to be a hot topic, judging from the forum activity. On the other hand, "Game development" is sexy

I put academia on purpose on the bottom to avoid the impression it's an "academic language". Maybe we can create a new grouping?
We could split academia into research and teaching, but then teaching would be alone.
Has someone else a better idea?

> This one is still broken, need to use LINK2 as below...

Rebased & fixed.

**Comment by jmh530** [comment]
The spacing on GPU programming looks wonky due to it being justified.

Also, I wouldn't necessarily group Numerical Computing, GPU Computing, and Data Science in to Academia. Maybe split them in to their own category. I might also mention how chaining range operations is like dplyr or something like that in the data science section.

File: areas-of-d-usage.dd [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            $(LI $(RELATIVE_LINK2 numerical, Numerical computing))
+            $(LI $(RELATIVE_LINK2 gpu, GPU programming))
+            $(LI $(RELATIVE_LINK2 data_science, Data science))
+            $(LI $(RELATIVE_LINK2 teaching, Teaching))
+        )
+    )
+)
+
+This list highlights a selection of great applications written in D
+from various areas. $(BR)
```
Comment (aG0aep6G) [inline areas-of-d-usage.dd]: I dislike these `$(BR)`s throughout. I'd rather see paragraphs wrapped in `$(P ...)`. In cases where that adds to much space, I'd say just join the lines, as there is no need for separation then.

**Comment by aG0aep6G** [comment]
This should be linked from somewhere. Documentation→Articles? Resources? I see that overview.html is apparently only linked from the Learn tour. That's not exactly great.

**Comment by wilzbach** [comment]
>The spacing on GPU programming looks wonky due to it being justified.

> Also, I wouldn't necessarily group Numerical Computing, GPU Computing, and Data Science in to Academia. Maybe split them in to their own category.

It's now named cutting-edge research.

> I might also mention how chaining range operations is like dplyr or something like that in the data science section.

On it - will come soon. Thanks for the idea.

> This should be linked from somewhere. 

I know added overview and this article to the bottom of dlang.org - I think it fits, because the heading is "Why D?". However the wording isn't the best yet.

> I see that overview.html is apparently only linked from the Learn tour. That's not exactly great.

I know (I made many PRs to the tour) - we are highly "underlinked"!

> Resources?

We have to be careful the current number of menu items is already too much.
Can we somehow create a new group?

> Articles

Could be an idea, but probably not the first place I search for an overview.

> Documentation.

We should merge http://dlang.org/comparison.html (linked from documentation) with http://dlang.org/overview.html

**Comment by wilzbach** [comment]
I had another look through the text and fixed a couple of points.

Btw that's how I propose to list the overview on the front page:

image

How should we proceed? More nitpicks?

**Comment by wilzbach** [comment]
Hey I went through the text again (minor revisions), but I added fancy icons - they are not optimal yet. The usual CSS hacks (responsive, nice alignment) are missing, however I think you could already check whether some icons don't fit

1) overview

image

2) sections

image

See [all](http://dtest.thecybershadow.net/artifact/website-fd5e6b914ec31f9e01b72737bc74f26d15dd8786-fa58e6418d1c0a99720e7479fc8c8d17/web/areas-of-d-usage.html)

**Comment by CyberShadow** [comment]
This looks really cool but I think should be approved by W/A. Normally this would probably belong on the wiki, but I really like how it's laid out.

* Some weird wrapping here, maybe just use a constant margin for the whole section:



* The icon currently used for "Industry" seems more associated with academy, I think.

File: areas-of-d-usage.dd [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    of D among game developers.
+    For example have a look at the cross-platform 3D game engine
+    $(HTTP dash.circularstudios.com, Dash) or the D game development toolkit
+    $(HTTPS github.com/d-gamedev-team/gfm, gfm).
+)
+
+$(SECTION3 $(LNAME2 web, Web applications), $(ARTICLE_FA_ICON cloud)
+
+    D's ability to combine low and high-level code in one language, asynchronous
+    features like Fibers and compile-time function execution makes it well suited
```
Comment (JackStouffer) [inline areas-of-d-usage.dd]: I would say "Fibers (a.k.a coroutines)"

Comment (ZombineDev) [inline areas-of-d-usage.dd]: Well according to [N4024](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4024.pdf) we have both - `core.thread.Fiber` looks like `boost::coroutine` and **vibe.d**'s use of fibers is more like `boost::fiber`.

So I would say "stack-full coroutines and fibers".

**Comment by wilzbach** [comment]
Last week I tried to make the icons more responsive. It's not perfect, but should look a lot better than before.
Moreover I add two nitpick runs over the text - can someone take 5-10 minutes to give the text a final nitpick round, s.t. we can ship an initial version? :)

**Comment by JackStouffer** [comment]
I think using [this](http://fontawesome.io/icon/industry/) for the industry icon would make more sense.

**Comment by JackStouffer** [comment]
Why is the industry icon on the other side of the page?
screen shot 2016-06-15 at 9 29 06 am

**Comment by JackStouffer** [comment]
I would remove the word "perfect" from the industry section. Your aim as a programmer is never to make perfect code as you know that's impossible.

**Comment by wilzbach** [comment]
> I think using this for the industry icon would make more sense.

It isn't part of FontAwesome 4.2  (see also #1371) - I guess to whether icon we decide there, I will just use the same here.

> Why is the industry icon on the other side of the page?

The idea was to have indent the subcategories, so I put for the categories the icons on the right side.
Is this too weird? Better ideas?

**Comment by JackStouffer** [comment]
"few scientific programmers care about" sounds negative. "few scientific programs need to worry about" sounds better IMO.

Also I would mention parallelism somewhere in that section.

**Comment by JackStouffer** [comment]
"Last but least" -> "Last but not least"

**Comment by JackStouffer** [comment]
I would link to the orgs using D page somewhere in the industry section

**Comment by JackStouffer** [comment]
> Is this too weird? Better ideas?

Yeah, it's kind of jarring. This effect is already achieved via the typography of the section headers.

**Comment by wilzbach** [comment]
> Yeah, it's kind of jarring. This effect is already achieved via the typography of the section headers.

Alrighty :)

> Also I would mention parallelism somewhere in that section.

I added:

"    allows to easily parallelize your algorithm and pipelines,"

> I would link to the orgs using D page somewhere in the industry section

I added:

"D has been used in many, diverse domains. A short selection will be presented below.
 For a full overview you can browse the list of reported $(LINK2 $(ROOT_DIR)orgs-using-d.html, organizations which use the D Language).

Thanks @JackStouffer :)
Do we need more eyes or is this ready for the first version?

**Comment by andralex** [comment]
I guess we're ready. Should we take away the funny but aged D mascot?

**Comment by JackStouffer** [comment]
Please squash commits

**Comment by wilzbach** [comment]
> I guess we're ready. Should we take away the funny but aged D mascot?

There is a related discussion for the DLang Tour:

https://github.com/stonemaster/dlang-tour/pull/248

> Please squash commits

With pleasure :)

**Comment by andralex** [comment]
Engage!

**Pull request merged**